### PR TITLE
WIP multiple IDS

### PR DESCRIPTION
**Pull request by rreinhardt9**
This is something we were pondering before in [this issue](https://github.com/lessonly/demux/issues/5)

### CODE REVIEW

**Comment by rreinhardt9** [comment]
Here are my thoughts after doing this spike...

TL;DR: I'm not sure that it's work the complexity introduced to accept "arrays of IDs" so that signals can have collections of object. Instead, I think we should focus signals on staying light and emphasis retrieving additional information through an API.

## What were we trying so solve?

We have a situation where you can make assignments in bulk. You can assign X number of users X pieces of content. Signals are designed to map to an individual object, but if we start thinking about scaling up that signal what do we do when we assign 40,000 users 10 pieces of content? Do we send 400,000 signals to each of the apps that are listening? If 10 apps are listening, that would mean then sending 4,000,000 signals 😱. While in most cases sending a signal for changes to an individual record makes sense and would not be a performance issue, the fact that we allow this bulk action in our app will require us to make sure the signal can ingest that as performantly as possible.

### All in one signal
Our first naive attempt was to capture the whole bulk assignment event into one signal sent at the time of assignment. Unfortunately, that meant that the time it took to transmit that signal increased linearly by "users X content" until it would fail because it took more than the 10 second timeout to transmit.

### Bulk Event ID
Our next thought was to have a single signal that sent an ID identifying the bulk assignment event. The partner app would then have to reach back to the API and try to figure out which of it's users (if any) where a part of that bulk assignment event.

This just moved the inefficiency, because now for every assignment event (even if it was for a single user that didn't use a specific partner app) each partner app would have to reach back to the API and inquire about every user it knew about to see if they were a part of that bulk event!!! We've just shifted the work from sending a big signal to lot's of API traffic. While at least we have ways of dealing with that (API limits, paging) it's still not efficient.

### Signal With Content per "actor" User

We've since added the ability for apps to listen for user specific signals. Now, it's possible for us to have a signal that would send a transmission to partner apps only for specific users that have installed the app.

We could now send a transmission only for users connected to specific apps and containing their assignment content details. This is a step in the right direction and gets us closer to a performant solution. It was this stage that started the conversation around if we should allow signals to take a collection; something like:
```
AssignmentSignal.new([1,3,5], account_id: 34).assigned
```
where `[1,3,5]` are the id's of related assignments.

This would then start to change the idea of a single "object" in the signals and a single integer for "object_id" in the transmission records.

In benchmarking, as we increased the number of assignments per user (keeping the number of users constant) we saw an increase in time required but it was nowhere near linear. When we increased the pieces of content 50 times, we only saw a doubling in the time. Still, it would be desirable to find a way to make this even more efficient by reducing the payload so that increasing the content did not increase the time needed to send the signal.

### Bulk Event ID per "actor" User

What we could do is send a signal that contains only the ephemeral information around a bulk event (maybe a message for example) and an ID of the bulk event. The partner app will only receive this signal for users that have set up the app and will know that it means that user was a part of that bulk event. The signal itself will be fast because it contains only basic information and that bulk event ID.
```
AssignmentSignal.new(nil, account_id: 34).assigned(bulk_assignment_id: 283strs8s9st)
```

We expect from benchmarking that this would mean a near linear increase in time to process that increases based on the number of users only. This could be a further refinement of the last situation; and instead of making a signal accept multiple IDs for an object the IDs of assignments could simply be context for now
```
AssignmentSignal.new(nil, account_id: 34).assigned(assignment_ids: [1,3,5])
```

This seems like the best compromise so far

**Comment by rreinhardt9** [comment]
Closing this for now as we are leaning in favor of not introducing this currently.
